k2/csrc/connect.cu

```diff
@@ -113,10 +115,14 @@ class Connector {
                   fsas_idx01x = fsas_row_splits2_data[fsas_idx01],
                   fsas_idx012 = fsas_idx01x + arcs_idx2,
                   fsas_dest_state_idx01 = dest_states_data[fsas_idx012];
-          // if this arc is a self-loop, just ignore this arc as we won't
-          // processe the dest_state (current state) again
+          // 1. If this arc is a self-loop, just ignore this arc as we won't
+          // processe the dest_state (current state) again.
+          // 2. If the state this arc pointing to is accessible, skip it.
+          // 3. If more than one arc leads to the same state, we select only
+          // one arc arbitrarily.
           if (fsas_dest_state_idx01 == fsas_idx01 ||
-              accessible_data[fsas_dest_state_idx01]) {
+              accessible_data[fsas_dest_state_idx01] ||
+              AtomicAdd(visited_data + fsas_dest_state_idx01, 1)) {
```

Comment:
Can you use
https://github.com/k2-fsa/k2/blob/9207b792ef1c25d112e3926212bddf9742514aca/k2/csrc/hash.h#L41

It would be more efficient if `int8` rather than `int32` is used to indicate a flag.

Reply:
I think `AtomicCAS` is not the one I want.  `AtomicAdd` here to confirm adding this state only once,  it would be called for many times and the return values are `0, 1, 2, 3, 4...`,  we add this state only when the return value is 0.

Reply:
I tried to look for one [here](https://docs.nvidia.com/cuda/cuda-c-programming-guide/#atomic-functions) and did not find any atomic function in `int8`.

Reply:
> we add this state only when the return value is 0.

Why does atomic compare and swap not work here? If I understand correctly, all you need is to know
whether the flag is 0 or not. If it is 0, then set it to one and keep the arc.

---

I just checked there is a version for int16 if you use atomicCAS.

Reply:
Yes, I made a mistake, `atomicCAS` works here.